### ISSUE TITLE
Remove Terminated

### ISSUE THREAD

**Issue by djspiewak** [scalaz/scalaz-stream#369]
People should not be throwing `Terminated(End)` to halt repetitions.  This particular semantic quirk is what makes the following two (seemly-identical) terms very distinct:

```scala
lazy val repeat1: Process[F, A] = await(t) { e => emit(e) ++ sound }

lazy val repeat2: Process[F, A] = await(t) { e => emit(e) } ++ sound
```

Oh hey, that's exactly the transformation that `pipe` and `tee` *assume* is sound!  What a coincidence…

In the above, `repeat2` will infinitely loop if the termination condition had previously been `t` throwing `Terminated(End)`, whereas `repeat1` will halt as expected with `End`.  I would humbly propose that this is insane, and we shouldn't be generating semantic quirks like this.

`Terminated` currently accepts any `Cause`, including `End`.  I would propose that, instead of this, we force `Terminated` to take only `EarlyCause`.  This makes the semantics a LOT cleaner (at compile time!) and forces users to explicitly write `onHalt` handlers to transmute thrown failures of this sort into normal stream completion (e.g. `wye`, `Queue` and `Topic` would all need such handlers).

tl;dr: Throwing `Terminated(End)` is deprecated behavior anyway, and retaining support for it is pushing us into highly unfortunate and non-composable semantic corners.  It should be removed.

**Comment by djspiewak** [comment]
Here's an example of a (failing) test which demonstrates how `Terminated(End)` breaks compositionality:

```scala
  property("identity piping preserve eval termination semantics") = secure {
    implicit val teq = Equal.equalA[Throwable]

    val halt = eval(Task delay { throw Terminated(End) }).repeat ++ emit(())

    ((halt pipe process1.id).runLog timed 3000 map { _.toList }).attempt.run === (halt.runLog timed 3000 map { _.toList }).attempt.run
  }
```

**Comment by pchlupacek** [comment]
@djspiewak `Terminated(End)` was just introduced to support signalling from the sink evaluation, namelly, io.*, that previsouly was implemented with just `throw(End)` to signal exhaust of the input source (essentially signal End). 

We anyhow need to touch all the `io` stuff with #333  change to assure proper release of the resources. However I didn't thought how to solve this. I am very much in favor of removing `Terminated` alltogether, if we ave good replacement.

**Comment by pchiusano** [comment]
I'm actually in favor of just getting rid of Terminated altogether. It is
never necessary - just use Option / Either and explicit control flow.

**Comment by djspiewak** [comment]
> I'm actually in favor of just getting rid of Terminated altogether. It is never necessary - just use Option / Either and explicit control flow.

I strongly agree.  Even in cases where boxing into `Either`/`Option` is prohibitively expensive, you can still throw a custom exception type and handle it later.  In other words, we're not *closing off* the efficient control flow path, we're just making it clear that it shouldn't be anyone's first choice.

**Comment by jedws** [comment]
if you _need_ an unboxed maybe type, consider https://gist.github.com/noelwelsh/02b9ecf6d21c7b8934e8

**Comment by djspiewak** [comment]
@jedws That's an amazingly clever trick, and probably something that we should use here.  It doesn't resolve *all* of the problems I am running into, removing `Terminated`, but it certainly addresses the performance warts.

**Comment by djspiewak** [comment]
After some work on this, I really don't think we can outright *remove* `Terminated` without some significant refactoring of public interfaces, particularly in the `async` package.  I'd rather not do something quite that dramatic, particularly since this is sort of blocking a more serious bug fix (#353).  So I would like to make the following proposal…

We keep `Terminated` as a standard "lifting" exception, but we remove the special handling from `eval`.  This fixes compositionality, since it forces users to implement their own `onHalt` when they want to perform the unlifting.  As with actual exceptions, *throwing* an exception is pure, it's only catching the exception that is impure and breaks composition.  Thus, we allow users to throw and catch all they want, but we don't automagically do the catching for them.  If users want to break composition, we're not going to hold their hand.

Seem reasonable?  This avoids changing `async` or `wye` dramatically, and it keeps an (unfortunately useful) primitive intact for the IO code.

**Comment by jedws** [comment]
At the risk of derailing the conversation…

> As with actual exceptions, throwing an exception is pure, it's only catching the exception that is impure and breaks composition.

How do I capture the value "throw exception"? For something to be pure I need to be able to replace the call with its value, but I cannot do that if I have thrown an exception, as there is no value to replace its evaluation with.

**Comment by djspiewak** [comment]
@jedws You can view `throw exception` as itself being a value (of type `Nothing`).  It's a lie, to be sure, but not a useful lie unless you can catch the exception.  If you never have any ability to catch exceptions (and you operate in a single thread of execution), then the `throw exception` value simply drops an atom bomb on your program, for all terms "exception".

**Comment by jedws** [comment]
@djspiewak the same argument applies to non-termination. IOW I do believe you just said there are no values of bottom, only nuclear explosions where programs used to exist :-)

**Comment by djspiewak** [comment]
@jedws Precisely so.  :-)  Non-termination can similarly be viewed as a value.

Of course, all of this is just meddling on the verge of insanity.  Referential transparency is merely a dying man's final, grasping attempt at sanity in non-total languages, and here you are stomping on the poor guy's fingers.

**Comment by jedws** [comment]
@djspiewak so, with subtyping and the fact that Nothing is a subtype of all types, your definition of non-totality as a value means that non-total == total! Stupendous! If by stomping on fingers I am rejecting the idea that we can in fact define non-totality as totality by rhetorical trick, then I stomp with force and conviction!

**Comment by djspiewak** [comment]
@jedws We're really just arguing over definitions here.  We don't lose a distinction between totality and non-totality if we define totality to require the absence of a bottom value.  There are some nice advantages to this definitional trick, such as the allowance of exceptions (but not exception handlers!).  Another advantage is we get to use Curry-Howard without waving hands vigorously.  (also, excitingly enough, there are elimination rules for bottom in the form of sum types, so we can derive calculi where bottom is a value, CHI holds and we get strictly better expressiveness than if we just say that bottom destroys CHI)

**Comment by jedws** [comment]
> We're really just arguing over definitions here. 

well, yes :-) I think they're important (and interesting) though, but not so we should necessarily be discussing here… I'll suspend and hopefully resume the conversation in a more suitable venue.

**Comment by pchlupacek** [comment]
+1 @djspiewak @jedws to any change here, unless this will break current code (source-wise) :-) I remember that terminated was introduced just because we didn't had anything better (I didn't like it either), and lets revisit that laters....

**Comment by fthomas** [comment]
We can get rid of `Terminated` in two places in `io` with a new `repeatEval` that takes a `F[Option[O]]` instead of a `F[O]`:  
```diff
diff --git a/src/main/scala/scalaz/stream/Process.scala b/src/main/scala/scalaz/stream/Process.scala
index ac7b745..6f3ec04 100644
--- a/src/main/scala/scalaz/stream/Process.scala
+++ b/src/main/scala/scalaz/stream/Process.scala
@@ -1436,6 +1436,9 @@ object Process extends ProcessInstances {
   def repeatEval[F[_], O](f: F[O]): Process[F, O] =
     awaitOr(f)(_.asHalt)(o => emit(o) ++ repeatEval(f))
 
+  def repeatEvalOption[F[_], O](f: F[Option[O]]): Process[F, O] =
+    await(f)(_.fold(empty[F, O])(o => emit(o) ++ repeatEvalOption(f)))
+
   /**
    * Produce `p` lazily. Useful if producing the process involves allocation of
    * some local mutable resource we want to ensure is freshly allocated
diff --git a/src/main/scala/scalaz/stream/io.scala b/src/main/scala/scalaz/stream/io.scala
index 07ac55d..6cf23bf 100644
--- a/src/main/scala/scalaz/stream/io.scala
+++ b/src/main/scala/scalaz/stream/io.scala
@@ -113,9 +113,9 @@ object io {
    * when processing the stream of lines is finished.
    */
   def linesR(src: => Source): Process[Task,String] =
-    resource(Task.delay(src))(src => Task.delay(src.close)) { src =>
+    resourceOption(Task.delay(src))(src => Task.delay(src.close)) { src =>
       lazy val lines = src.getLines // A stateful iterator
-      Task.delay { if (lines.hasNext) lines.next else throw Cause.Terminated(Cause.End) }
+      Task.delay { if (lines.hasNext) Some(lines.next) else None }
     }
 
   /**
@@ -154,6 +154,13 @@ object io {
       repeatEval(step(r)).onComplete(eval_(release(r)))
     }
 
+  def resourceOption[F[_],R,O](acquire: F[R])(
+                               release: R => F[Unit])(
+                               step: R => F[Option[O]]): Process[F,O] =
+    await(acquire) { r =>
+      repeatEvalOption(step(r)).onComplete(eval_(release(r)))
+    }
+
   /**
    * The standard input stream, as `Process`. This `Process` repeatedly awaits
    * and emits chunks of bytes  from standard input.
@@ -166,7 +173,7 @@ object io {
    * and emits lines from standard input.
    */
   def stdInLines: Process[Task,String] =
-    Process.repeatEval(Task.delay { Option(scala.Console.readLine()).getOrElse(throw Cause.Terminated(Cause.End)) })
+    Process.repeatEvalOption(Task.delay(Option(scala.Console.readLine())))

```
Any reason not to do this?

**Comment by pchlupacek** [comment]
@fthomas how about instead of proposed resourceOption signature having just process signature instead: 

```scala

def resource[F[_],R,O](acquire: F[R])(
release: R => F[Unit])(
step: R => F[Process[F,O]]): Process[F,O] = ???


```

perhaps, the step can be only `step: R => Process[F,O]`, not sure on that tho. 

That way we can also signal correct reason for termination, that is not possible with Option[O] method.

**Comment by djspiewak** [comment]
@fthomas @pchlupacek There is a bigger problem with getting rid of `Terminated`, which is `through`.  The problem is that it is *very* common for a `Sink` to need to signal termination back upstream depending on the values it consumes.  This is not really something that can be resolved directly with any sort of `Option` scheme unless we change `Sink` to be `Process[F, A => F[Option[Unit]]]`, which is both clumsy and incompatible.

Ultimately, the most flexible approach would be to figure out how to make transducers functional (of the style shown in #351) and replace `Sink` entirely, since effectful transducers give us the back-signal control of `Process1`.  Unfortunately, that's not an easy change at all.

In the meantime, I'm about halfway through a compromise change, which is to say removing the auto-magical translation of `Terminated` causes in `eval` and `repeatEval` while leaving `Terminated` and `asHalt`.  I'm switching to `Option` results in as many places as I can (e.g. `io.stdIn`), and just manually inserting `asHalt` wherever I can't.

**Comment by pchlupacek** [comment]
@djspiewak  still don't see why we cannot define a step like ` R => Process[F,O]` ? That will allow us to signal whatever we want to upstream? And we can then only define resource as 
```scala

def resource[F[_],R,O](acquire: F[R])(
                         release: R => Process[F,Nothing])(
                         sf: R => Process[F,O]): Process[F,O] =
    await(acquire) { r =>
      def go() = {
      sf(r).step {
        case Step(Await(), cont) => // attach next `go`
        case Step(Emit(),cont) => //attach next `go`
        case Halt(rsn) => release(r)
      }
      }
      go()
    }


```

You are right, tho that we need to somehow solve `eval` which is now extracting the `Terminated`. I am not sure tho how often you want really signal terminated `from inside` of the channel, and if so, perhaps correct way to code it to return from the channel the Process[F,O] instead of throwing and exception in Task. 

I agree that solving this perhaps requires touch to channel/sink signature, and as already proposed, I think we would be completely ok if sink would be:
```scala
type Sink[+F[_],-O] = Process[F, O => F[Process[F,Nothing]]] 
type Channel[+F[_],-I,O] = Process[F, I => F[Process[F,O]]]

//Perhaps we can even discard the `F` in function

type Sink[+F[_],-O] = Process[F, O => Process[F,Nothing]]
type Channel[+F[_],-I,O] = Process[F, I => Process[F,O]]
```

**Comment by djspiewak** [comment]
@pchlupacek I'm not sure I see what you're proposing with `step`.  What problem does the signature in question solve?  I really don't understand your example.

Regarding the proposed `Sink` and `Channel` signatures, you'll notice that neither of those signatures allow signaling to `through` that the channel has determined *effectfully* that it is time to terminate.  Again, this is a very common use-case, even within scalaz-stream itself.  Basically, this is a poor-man's variant of having a full effectful state machine inside of `Channel`, which is a universally useful generalization that we do not currently support.

**Comment by pchlupacek** [comment]
@djspiewak I am not sure what do you mean that channel has determined effectfully to terminate. 
Channel, is Process right? and If that process terminates then the whole process terminates. I think this is the behaviour we have now. 

I think you have said that in current scenario R => Task[O] signature requires us to throw Terminated(cause) to essentially signal termination. However with proposed change you just return that cause, or you just return await() producing that cause on its evaluation, eventually. 

Additionally this signature for channel allows you to return one, none or many values at once, which current signature does not.  

As on the example, by stepping through the process, we don't need any Terminated hacks, we just do resource management directly here.

**Comment by djspiewak** [comment]
@pchlupacek 

> I am not sure what do you mean that channel has determined effectfully to terminate. 
Channel, is Process right? and If that process terminates then the whole process terminates. I think this is the behaviour we have now.

It is not.  Consider the following semantic for a sink:

1. For each integer value
2. Compute a random number, weighted by the incoming value
3. Round random number to the nearest integer
4. If the rounded value is even, terminate, otherwise emit the output

This is easy to do with transducers (see #351), but it is impossible to do cleanly with the current `Channel` encoding.  The only way to implement this right now is to throw `Terminated`, because the determination of whether or not to terminate is *inside* the effect, rather than as part of the outer structure of the process.  In other words, it doesn't help us that `Channel` is a `Process`, because there are precious few use-cases for `Channel` that are *not* just `channel.lift(f)`, for some function `f` (the use cases which are less trivial than that all boil down to nondeterministic multiplexing, which is a pretty specific use-case when you think about it).

The new proposed encoding for `Sink` does not resolve the problem I illustrate above.

> think you have said that in current scenario R => Task[O] signature requires us to throw Terminated(cause) to essentially signal termination. However with proposed change you just return that cause, or you just return await() producing that cause on its evaluation, eventually.

In the proposed encoding, `Sink = Process[F, A => Process[F, Nothing]]`.  You can't "return" anything.  The inner process always produces a `Halt`.  You can produce a `Halt(Error(Terminated(End))))` if you really want to, but is that *at all* superior to the current `throw Terminated` situation?

> Additionally this signature for channel allows you to return one, none or many values at once, which current signature does not.

It does, you just need to flatten afterward.  (starting from `Channel[F, A, Process[F, B]]`).

**Comment by pchlupacek** [comment]
@djspiewak I am perhaps missing something but can't just   ```Halt(Error(Terminated(End))))``` be replaced as simple ```Kill``` ?

As for the `Sink`, I agree that perhaps is good to keep `Sink` returning `Unit`, as you can always drain it anyhow. The ```Process[F, A => Process[F, Unit]]``` will just do the job as it does today. Although, the same can be achieved with observe, plus instead of nothing telling `Unit` you will get element passed to `Sink`. 

As for the flattening, sure this pollutes our code a lot :-). But I think the real good is to reuse process signalling for termination.

**Comment by pchlupacek** [comment]
@djspiewak  just for clarification the example with proposed Sink/Channel signature will be as : 

```scala

 val integers:Process[Task,Int] = Process.range(0,Int.MaxVaule)

 val randomChannel : Channel[Task,Int,Int] = constant {
    (incoming:Int) => 
       eval(Task.delay(Random.nextInt(incoming)))
       .flatMap {
          case i if i % 2 == 0 => emit(i)
          case i => Kill
        }
  }

```

**Comment by djspiewak** [comment]
@pchlupacek We cannot simply use `Kill` because of the following possibility:

```scala
(emit(true) wye (src to sink append src2))(wye.interrupt)
```

The `Kill` will be raised potentially during execution of the flattened sub-process within the sink.  This is a problem, because some code in `to` needs to manage converting `Kill` back into `End` within the outer process.  This of course means that kills raised at the "wrong" time could be completely swallowed by appends, which is what we see with `src2`.

Put more concretely: if we use `Kill` as a signal for a sink to terminate the outer fused process, then the above will *nondeterministically* emit `src2`, depending on thread scheduling.

**Comment by pchlupacek** [comment]
@djspiewak hence `to` is in fact `tee`, and I think we already perform the `Kill` swallowing in `tee`, so don't think this will be an issue. I think we can just reuse that and convert any Killed process on `tee` to be normally terminated.

**Comment by pchlupacek** [comment]
put it in other way, I don't see any reason why we simply cannot do 
```scala
(source to sink).swallowKill
```
Behind the scenes.

**Comment by djspiewak** [comment]
@pchlupacek Please see the following:

```scala
  property("kill appended chain") = secure {
    (0 until 100) forall { _ =>
      val q = async.unboundedQueue[Int]
      val data = emitAll(List(1, 2, 3))

      val results = (emit(true) wye (q.dequeue observe q.enqueue append data))(wye.interrupt).runLog.timed(3000).run

      results == Seq()
    }
  }
```

Swallowing `Kill` would most definitely be an issue.  Your proposal would cause the above test, which currently passes on master, to fail.

**Comment by pchlupacek** [comment]
@djspiewak how it comes that `q.enqueue` is evaluating to `Terminated(End)` in this example? I don't see that in your code, you closing the queue somehow?

**Comment by pchlupacek** [comment]
Ah you mean swallowing the `Kill`. That won't be an issue, hence you may distinguish if that kill comes from `inner` part of `to` combinator. That will be swallowed, rest will be in `tee` as is. So this will pass.

**Comment by djspiewak** [comment]
@pchlupacek When `wye` kills the the `tee`, one of the `Await`(s) inside of that `tee` will be "active" in `stepAsync`.  The results of this kill will be that the `Await` in question will produce a `Halt(Kill)`.  Note that this `Await` may very well be the `Await` inside of the `Sink`.  When this happens, the results are *indistinguishable* from when the `Await` inside of the `Sink` triggers premature termination in the outer process.  However, in order to avoid propagating `Kill`s unnecessarily, we need to swallow premature termination kills.  It is this kill swallowing which will cause the test I gave to non-deterministically break under your proposal.

**Comment by pchlupacek** [comment]
consider `through` be denfined as :
```scala 
type Channel[+F[_],-I,O] = Process[F, I => Process[F,O]]
def through[F2[x]>:F[x],O2](f: Channel[F2,O,O2]): Process[F2,O2] = {
  self.zipWith(f)((o,f) => f(o)).flatMap(identity)
}

```
Do we need `swallowKill` at all?

**Comment by djspiewak** [comment]
@pchlupacek Kills propagate out from `Await`s inside of `tee`.  If they didn't, my test above would fail.  So yes, in order to make your proposal work, we would need `swallowKill`.  Probably in the following way:

```scala
def through[F2[x]>:F[x],O2](f: Channel[F2,O,O2]): Process[F2,O2] = {
  self.zipWith(f)((o,f) => f(o).swallowKill).flatMap(identity)
}
```

However, if we do that, my test will start nondeterministically failing.

**Comment by pchlupacek** [comment]
@djspiewak you have been posting faster than me :-) I think I see that issue, let me think...

**Comment by pchlupacek** [comment]
@djspiewak you right we need to keep state of the `Kill` somehow, obviously this is doable, and I think still better than current `Terminated` solution, but perhaps not that good as we would like it to be:

```scala

def through[F2[x]>:F[x],O2](f: Channel[F2,O,O2]): Process[F2,O2] = suspend {
  var normally:Boolean = false
  self.zipWith(f)((o,f) => f(o)).flatMap(p => p.onKill(eval_(Task.delay(normally = true))))
 .onHalt {
   case Kill if normally => halt
   case cause => cause 
  }
}

```

**Comment by djspiewak** [comment]
@pchlupacek There are two problems with that proposed implementation.  First, it doesn't compile.  The `through` combinator purports to allow arbitrary effects, `F2`, but your implementation requires that only `Task` effects are present.

The second problem though is that it won't work.  It's basically doing the same thing as `swallowKill` would.  The `onKill` handler is going to get triggered regardless of whether the `Kill` came from the `Channel` or if it came from some enclosing concurrency interpreter.

**Comment by pchlupacek** [comment]
I think the onkill is wrapped in inner process. Then you will distinguish the two scenarios. Sorry for this not to compile. I am writing this on mobile. Instead of task we can use suspend. That will allow any 'F'

**Comment by djspiewak** [comment]
@pchlupacek You cannot distinguish between the two scenarios because even the inner process can and will be killed by an outer `stepAsync`.  That's basically the whole point of my example.  If the inner process were *not* getting killed, then my example would hang.

**Comment by pchlupacek** [comment]
Well.  I think this is really solvable. Anyhow lets merge your changes for now. And lets see if we can get better than that on transducers.

**Comment by djspiewak** [comment]
I'm going to call this "resolved" for now.  If we want to track the full removal of `Terminated`, we should open a new ticket.  Note that full removal is going to require not only transducers but also some significant refactoring of the IO and async tools.
